optbuilder: fix panic on upsert with concurrent ALTER ADD of virtual computed column

msirek opened the pull request:
A panic may happen when an upsert or other DML is run in one SQL
connection while simultaneously an ALTER TABLE ADD COLUMN of a virtual
computed column is run in a different SQL connection.

This is due to `Builder.buildScan` attempting to project the
newly-added virtual column as part of index arbiter processing
for the upsert, causing an access to the tabMeta.ComputedCols map,
which has not been allocated, resulting in a panic due to the nil-pointer
access. Other DML operations like updates, inserts and deletes may also
observe this behavior.

The fix is to collect the oridinals of all virtual computed columns
under mutation inside `buildScan`, from the set of column ordinals being
selected, and pass that to addComputedColsForTable. If the computed
column for which we're building the computed column expression is in
this `includeVirtualMutationColOrds` set, then always build the
expression, even when the column is under mutation.

Fixes https://github.com/cockroachlabs/support/issues/2457
Fixes #107357

Release note (bug fix): This patch fixes an internal error in UPDATE,
UPSERT, INSERT or DELETE statements run concurrently with
ALTER TABLE ADD COLUMN of a virtual computed column on the same table.

Co-authored-by: Marcus Gartner <marcus@cockroachlabs.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/106782)
<!-- Reviewable:end -->

michae2 reviewed (changes requested):
Nice job figuring this out so quickly!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_Kkcag1U48OTE-WV81:b-2be2dn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @msirek)

___
*[`pkg/sql/opt/optbuilder/select.go` line 565 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_KkRZW9iuAgfsfadQa:-N_KkRZW9iuAgfsfadQb:batj60q) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdd03831451904a6b02613eb94032967797fdb78/pkg/sql/opt/optbuilder/select.go#L565)):*
> ```Go
> 			typ:          col.DatumType(),
> 			visibility:   columnVisibility(col.Visibility()),
> 			kind:         kind,
> ```

It looks like we're adding other mutation columns to the `outScope` to sometimes use as fetch columns in `INSERT`, `UPDATE` and other mutation statements. I'm currently trying to figure out if we would ever need to add a virtual computed column as a fetch column. If so, then we might need to include the virtual computed column.


<!-- Sent from Reviewable.io -->

fqazi commented:
> Nice job figuring this out so quickly!
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_Kkcag1U48OTE-WV81:b-2be2dn)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @msirek)
> 
> _[`pkg/sql/opt/optbuilder/select.go` line 565 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_KkRZW9iuAgfsfadQa:-N_KkRZW9iuAgfsfadQb:batj60q) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdd03831451904a6b02613eb94032967797fdb78/pkg/sql/opt/optbuilder/select.go#L565)):_
> 
> > ```go
> > 			typ:          col.DatumType(),
> > 			visibility:   columnVisibility(col.Visibility()),
> > 			kind:         kind,
> > ```
> 
> It looks like we're adding other mutation columns to the `outScope` to sometimes use as fetch columns in `INSERT`, `UPDATE` and other mutation statements. I'm currently trying to figure out if we would ever need to add a virtual computed column as a fetch column. If so, then we might need to include the virtual computed column.

Shouldn't INSERT/UPDATE statements error out if a virtual column computation would be bad? For example, lead to a division by zero?

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @michae2)

___
*[`pkg/sql/opt/optbuilder/select.go` line 565 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_KkRZW9iuAgfsfadQa:-N_L3oWU02X0tUrS1PSI:b-29s0y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdd03831451904a6b02613eb94032967797fdb78/pkg/sql/opt/optbuilder/select.go#L565)):*
> I'm currently trying to figure out if we would ever need to add a virtual computed column as a fetch column. If so, then we might need to include the virtual computed column.

Yeah, let me look a little harder for such a case too. I originally wanted to always fill in computed column info, so it's always available, but that change ran into so many bugs, so I'm hoping it's not necessary.


<!-- Sent from Reviewable.io -->

michae2 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, and @msirek)

___
*[`pkg/sql/opt/optbuilder/select.go` line 565 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_KkRZW9iuAgfsfadQa:-N_L401b5eCteNpMDucl:b-4mpbyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdd03831451904a6b02613eb94032967797fdb78/pkg/sql/opt/optbuilder/select.go#L565)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

> I'm currently trying to figure out if we would ever need to add a virtual computed column as a fetch column. If so, then we might need to include the virtual computed column.

Yeah, let me look a little harder for such a case too. I originally wanted to always fill in computed column info, so it's always available, but that change ran into so many bugs, so I'm hoping it's not necessary.
</blockquote></details>

Here's an example I'm thinking about:

```sql
CREATE TABLE t (a INT PRIMARY KEY);
INSERT INTO t VALUES (4);
SET CLUSTER SETTING jobs.debug.pausepoints = 'schemachanger.before.exec,newschemachanger.before.exec';
ALTER TABLE t ADD COLUMN b INT AS (a + 1) VIRTUAL UNIQUE;
EXPLAIN (OPT, VERBOSE) UPDATE t SET a = a + 10 WHERE a > 0;
```

If there were no schema change, I think `b` would need to be included as a fetch column in the `UPDATE` so that the update could `Del` the `b=5,a=4` row from the secondary index `t_b_key` and `Put` the `b=15,a=14` row. I don't understand schema changes well enough to be 100% certain, but I think we might need to do that for a mutation column, too, in case the row has already been backfilled but the schema change isn't done yet.


<!-- Sent from Reviewable.io -->

msirek commented:
Making some changes, hold off on reviewing further for now.

msirek reviewed (commented):
RFAL. Thanks for the example test case!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_U5nCcAj0PM1DRJSbT:bc6knds)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @michae2)

___
*[`pkg/sql/opt/optbuilder/select.go` line 565 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_KkRZW9iuAgfsfadQa:-N_L66Y--fSnmXl81U5D:b-9yf4mu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdd03831451904a6b02613eb94032967797fdb78/pkg/sql/opt/optbuilder/select.go#L565)):*
Thanks, I've modified the tests to include the UNIQUE keyword, which caused the tests to fail with the original reviewed version of the code.

The new version builds the virtual computed column expression when that column is under mutation (ADD/DROP), but only in calls to `buildScan` which specifically require the mutation columns. When I previously tried this fix I had built it in all cases, which caused many other ALTER TABLE cases to fail. So, restricting it to cases where we 

A virtual computed column can't be a referencing or referenced column in a foreign key constraint, so we don't need to worry about those paths. I believe handling cases where `columnKinds` of `includeMutations` == `true` should be exactly the cases we should handle, no more, no less.

> Shouldn't INSERT/UPDATE statements error out if a virtual column computation would be bad? For example, lead to a division by zero?

Yes, now the virtual computed column expression will be evaluated in a projection after the scan, and if there is such an error, the statement will error out.


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/optbuilder/select.go` line 864 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_ZTQFiBJ6V5AuHpFp3:-N_ZTQFiBJ6V5AuHpFp4:b14k0s6) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L864)):*
> ```Go
> 			// computed column expression value (in general). Virtual computed columns
> 			// may be needed as "fetch" columns in order to build values to remove
> 			// from unique indexes which include a virtual column as a key column.
> ```

This explains why the virtual columns are needed, but not why its valid to put them in the metadata. The computed columns in the metadata are "known truths" and therefore must be backfilled fully, otherwise the optimizer would make incorrect assumptions and produce incorrect query plans. But this case is confusing because virtual columns require no backfill, so are they always "known truths"?

I don't have a solid mental model of what it means for a virtual column to be in a mutation state. When would it go through this state and would "reading it" (producing its expression) ever produce an incorrect value for a row? Perhaps @fqazi could explain more from the schema side.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)

___
*[`pkg/sql/opt/optbuilder/select.go` line 864 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_ZTQFiBJ6V5AuHpFp3:-N_ZVuFB97k-NeYg-l6j:b-yxixx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L864)):*
> I don't have a solid mental model of what it means for a virtual column to be in a mutation state.

As I understand it, this happens when an ALTER TABLE ADD/DROP column modifies the table descriptor at the beginning of the transaction, while the ALTER TABLE is running as a job. When the job completes, the column in the descriptor is modified to remove the mutation state. Maybe there are more cases that can set a mutation state, I'm not sure. Someone please correct me if this is incorrect.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)

___
*[`pkg/sql/opt/optbuilder/select.go` line 864 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_ZTQFiBJ6V5AuHpFp3:-N_ZXCrYEFGoOcS2Ws9P:b2xa3q6) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L864)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

> I don't have a solid mental model of what it means for a virtual column to be in a mutation state.

As I understand it, this happens when an ALTER TABLE ADD/DROP column modifies the table descriptor at the beginning of the transaction, while the ALTER TABLE is running as a job. When the job completes, the column in the descriptor is modified to remove the mutation state. Maybe there are more cases that can set a mutation state, I'm not sure. Someone please correct me if this is incorrect.
</blockquote></details>

In the case of DROP column, the column entry would be removed from the descriptor when the job completes (or maybe there is some other marker, like when the backfill completes).


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)

___
*[`pkg/sql/opt/optbuilder/select.go` line 864 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_ZTQFiBJ6V5AuHpFp3:-N_Z_h8b3J8qTVl9scQA:bqu2we9) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L864)):*
> But this case is confusing because virtual columns require no backfill, so are they always "known truths"?

Probably their value is a known truth, but whether they should be included in the scan would be determined by the context of the scan. A scan which happens before the ALTER ADD COLUMN job is complete should not see the column, and so that why the call to `addComputedColsForTable` skips over it if it's marked as a mutation column. I think only a scan which should see the computed column (e.g. in order to check for duplicates, etc.) would have its own `TableMeta` and its own `TableMeta.ComputedCols` built, so other scans which shouldn't see the column, their `TableMeta` would not have an entry for that column in their `TableMeta.ComputedCols`.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
I believe this is related to #79613, #79691, and my attempted fix for these issues in https://github.com/mgartner/cockroach/commit/a16e03f43ec1b1c4c2035f20ba62a44b69c88245, which I never got fully working. The issue is that there is a contradiction: we need to produce the mutation virtual computed column in order to write to the mutation index, but we can't read the mutation virtual column because it's in the mutation state.

> This is due to `Builder.buildScan` attempting to project the  
newly-added virtual column as part of index arbiter processing  
for the upsert, causing an access to the tabMeta.ComputedCols map

Is this because the virtual column is part of a unique index? If not, why is the buildScan attempting to project it during index arbiter processing?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_Zdv2FEZxxp5Nb82QF:b-mu3sts)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @michae2, and @msirek)

___
*[`-- commits` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_ZdBvp4KOQjGKQX8rI):*
It's a nil-pointer exception, not a segfault, right?


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Also see #98692 which attempted to fix the same (and/or similar) issue(s).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N__7i1I6Jf-LsjvvT18:b-63jkig)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @michae2, and @msirek)



<!-- Sent from Reviewable.io -->

mgartner commented:
I think there's a larger question to answer here: Is it valid for an index and a column in the index to both be in mutation states at the same time? I've started a thread here to discuss with the foudations team: https://cockroachlabs.slack.com/archives/C0168LW5THS/p1689621357959539





<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @michae2, and @msirek)

___
*[`pkg/sql/opt/optbuilder/select.go` line 527 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N__eGanCoiDu-AQFcxG:-N__eGanCoiDu-AQFcxH:blp168) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L527)):*
> ```Go
> 	inScope *scope,
> 	disableNotVisibleIndex bool,
> 	includeVirtualMutationCols bool,
> ```

The usage of this field exactly matches the `columnKinds{includeMutations: ...}` field of the argument to `tableOrdinals`, and the set of virtual column projections are built from those ordinals. Instead of adding a boolean field here, we should accumulate the virtual column ordinals and pass those to `addComputedColsForTable` to determine which virtual computed columns to include. That way it's not possible for the ordinals and the extra boolean argument to be inconsistent.
___
*[`pkg/sql/opt/optbuilder/select.go` line 864 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_ZTQFiBJ6V5AuHpFp3:-N__bOFXE-V0qiHg2jsf:b-ei8wj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L864)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

> But this case is confusing because virtual columns require no backfill, so are they always "known truths"?

Probably their value is a known truth, but whether they should be included in the scan would be determined by the context of the scan. A scan which happens before the ALTER ADD COLUMN job is complete should not see the column, and so that why the call to `addComputedColsForTable` skips over it if it's marked as a mutation column. I think only a scan which should see the computed column (e.g. in order to check for duplicates, etc.) would have its own `TableMeta` and its own `TableMeta.ComputedCols` built, so other scans which shouldn't see the column, their `TableMeta` would not have an entry for that column in their `TableMeta.ComputedCols`.
</blockquote></details>

I think https://github.com/cockroachdb/cockroach/pull/60733 provides a crucial restriction that ensures that virtual computed column projections always produce a correct value during a backfill: they are not allowed to reference other write-only or delete-only columns. So I think they can be considered "known truths".

Does that sound correct to you? Can you think of any other cases where they wouldn't produce the correct value? If not, can you add this explanation and maybe a link to that PR 60733 to the comment here?


<!-- Sent from Reviewable.io -->

mgartner commented:
It would be great to also add some optimizer tests for these cases. You can simulate write-only and delete-only columns and indexes using a special naming scheme in optimizer tests. For example, something like this would be a good test to add to `pkg/sql/opt/optbuilder/testdata/virtual-columns`: 

```
exec-ddl
CREATE TABLE mut (
  a INT PRIMARY KEY,
  "b:delete-only" INT AS (a + 1) STORED,
  UNIQUE INDEX "v_idx:delete-only" (b)
)
----

opt
UPDATE mut SET a = a + 1
----
```





<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @msirek)

___
*[`pkg/sql/opt/optbuilder/select.go` line 527 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N__eGanCoiDu-AQFcxG:-N__syqu4jmzmlCutB0T:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L527)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

The usage of this field exactly matches the `columnKinds{includeMutations: ...}` field of the argument to `tableOrdinals`, and the set of virtual column projections are built from those ordinals. Instead of adding a boolean field here, we should accumulate the virtual column ordinals and pass those to `addComputedColsForTable` to determine which virtual computed columns to include. That way it's not possible for the ordinals and the extra boolean argument to be inconsistent.
</blockquote></details>

+1


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
>>This is due to `Builder.buildScan` attempting to project the  
newly-added virtual column as part of index arbiter processing  
for the upsert, causing an access to the tabMeta.ComputedCols map

> Is this because the virtual column is part of a unique index? If not, why is the buildScan attempting to project it during index arbiter processing?

Good question. My perspective on this is, whether the column is needed or not by a given processing flow, it has been requested, so it should be built. If it's sometimes being built when it's not needed, that seems like it would be a separate bug in the caller's logic that could be addressed independently. I think the virtual column is built even if there's no unique index. When there is a unique index, the fetch column is actually used to remove the old entry before the new one is written.

The tests in upsert_test.go do catch errors when the computed column is built incorrectly, so we do have test validation of this case.

> Is it valid for an index and a column in the index to both be in mutation states at the same time?

Maybe someone on @cockroachdb/sql-foundations could comment on that in this PR.

> It would be great to also add some optimizer tests for these cases.

Done

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_fxcuyAI3GkIhG7rjn:b-nw1lks)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)

___
*[`-- commits` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_ZdBvp4KOQjGKQX8rI):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It's a nil-pointer exception, not a segfault, right?
</blockquote></details>

Ya, I guess it's caught and we panic. Updated the text.
___
*[`pkg/sql/opt/optbuilder/select.go` line 527 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N__eGanCoiDu-AQFcxG:-N_ddlN5ElhBzIoX62p9:b3kjtuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L527)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

+1
</blockquote></details>

OK, I've updated the `tableOrdinals` function to also generate the slice of computed column ordinals to feed into `addComputedColsForTable`.
___
*[`pkg/sql/opt/optbuilder/select.go` line 864 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_ZTQFiBJ6V5AuHpFp3:-N_f8zCeAEbrt9v39QtE:b-huods1) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L864)):*
> I think [#60733](https://github.com/cockroachdb/cockroach/pull/60733) provides a crucial restriction that ensures that virtual computed column projections always produce a correct value during a backfill: they are not allowed to reference other write-only or delete-only columns. So I think they can be considered "known truths".

Yes, this seems crucial and probably protects not just the case we're handling here, but other problem cases too. I've updated the comment, which has been moved into the `tableOrdinals` function. I can't think of any other problem cases.


<!-- Sent from Reviewable.io -->

msirek commented:
With the latest push I'm getting some errors, so obviously there is a logic mismatch after moving the fix into `tableOrdinals`. Will look into it tomorrow.

msirek reviewed (commented):
RFAL
I found the regressions were caused by only producing computed column ordinals from the set of table ordinals. The fix was to match the old looping behavior in `addComputedColsForTable` and look at all columns in `TableMeta`. We may want to see if this could be changed in the future, but it's probably not best to make more changes in this area through this PR which might cause regressions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_j57TdDazjuopOQit9:b1zmd0e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> My perspective on this is, whether the column is needed or not by a given processing flow, it has been requested, so it should be built.

We have to be especially careful with the ComputedCols field of table metadata because the optimizer considers these to be known truths about the table, like a not-null constraint. If we add an expression that doesn't hold true for every row in the table, the optimizer could generate an incorrect query plan. So we can't blindly assume that we can add any expression to ComputedCols.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_k6Mqc6y43ArcYktVU:biagjgd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @michae2, and @msirek)

___
*[`pkg/sql/opt/optbuilder/select.go` line 527 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N__eGanCoiDu-AQFcxG:-N_k66q1AyEBhPIUySh-:bt0v2d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L527)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

OK, I've updated the `tableOrdinals` function to also generate the slice of computed column ordinals to feed into `addComputedColsForTable`.
</blockquote></details>

Can you build the set of virtual computed column ordinals in `buildScan` instead? That should be simpler.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
Yes, I agree. What I meant was, bugs dealing with improper building of `ComputedCols` is a bug in the caller and should be corrected on the caller's end, and the callee should simply be a simple function that does what it's told without any special logic inside of it to override the caller. I'm not trying to say it's correct for a caller to build something when it's not needed or when it's incorrect to build it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_m71vJ6Y0J5WqguA8z:b7dwbnh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)

___
*[`pkg/sql/opt/optbuilder/select.go` line 527 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N__eGanCoiDu-AQFcxG:-N_m9FXk2j4GWwIIVM6I:bata9lw) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L527)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you build the set of virtual computed column ordinals in `buildScan` instead? That should be simpler.
</blockquote></details>

If I do that, I will have to re-introduce the `includeVirtualMutationCols` parameter to `buildScan`, because we don't have the `includeMutations=true` information, needed to decide when to add virtual computed mutation columns to `TableMeta.ComputedCols` present within `buildScan`. Then the solution will look pretty much the same as the original fix, except the loop over all `cat.Table.Column`s is just hoisted a single level out of `addComputedColsForTable`, to produce an ordinal list one line before calling that function, which uses it. It seems a little strange to do the same exact loop, just one step out of the function and one line before it, instead of within the function:
```go
	for i := 0; i < n; i++ {
		col := tab.Column(i)
                if (condition_expression_involving_includeVirtualMutationCols) {
                   // add i to computedColOrdinals list
                }
        }
        b.addComputedColsForTable(tabMeta, computedColOrdinals)
```
I'll make the change if you think it's best, but I thought the point of the previous change request was to do away with the `includeVirtualMutationCols` parameter.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)

___
*[`pkg/sql/opt/optbuilder/select.go` line 527 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N__eGanCoiDu-AQFcxG:-N_nq0P_4jbCo1wgNcB9:b-349b2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/65132b85c336fe895f3f69a77570560db0951141/pkg/sql/opt/optbuilder/select.go#L527)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

`ordinals` already tells us whether we need to build a virtual mutation column or not. It will only contain the ordinal of a virtual mutation column if `tableOrdinals` was called with `includeMutations: true`. Here's a rough sketch of what I had in mind: https://github.com/mgartner/cockroach/commit/3723ec0613c1a4a530231609db5ebabe10015039
</blockquote></details>

That would work if it's OK to limit the columns to those already in the `ordinals`, but when I tried that fix previously there were a ton of ALTER TABLE bugs.  So, the current assumptions made elsewhere in the code must be that we loop through all columns in the table when building computed column expressions, no matter what's in the ordinals. Some of these assumptions may be for hash sharded indexes, which may need the virtual column hash bucket for the row, or maybe for other features that examine the `ComputedCols` map without regard for the actual selection columns.  


<!-- Sent from Reviewable.io -->

Commit:
optbuilder: fix panic on upsert with concurrent ALTER ADD of virtual computed column

A panic may happen when an upsert or other DML is run in one SQL
connection while simultaneously an ALTER TABLE ADD COLUMN of a virtual
computed column is run in a different SQL connection.

This is due to `Builder.buildScan` attempting to project the
newly-added virtual column as part of index arbiter processing
for the upsert, causing an access to the tabMeta.ComputedCols map,
which has not been allocated, resulting in a panic due to the nil-pointer
access. Other DML operations like updates, inserts and deletes may also
observe this behavior.

The fix is to collect the oridinals of all virtual computed columns
under mutation inside `buildScan`, from the set of column ordinals being
selected, and pass that to addComputedColsForTable. If the computed
column for which we're building the computed column expression is in
this `includeVirtualMutationColOrds` set, then always build the
expression, even when the column is under mutation.

Fixes https://github.com/cockroachlabs/support/issues/2457
Fixes #107357

Release note (bug fix): This patch fixes an internal error in UPDATE,
UPSERT, INSERT or DELETE statements run concurrently with
ALTER TABLE ADD COLUMN of a virtual computed column on the same table.

Co-authored-by: Marcus Gartner <marcus@cockroachlabs.com>

msirek reviewed (commented):
Don't recall. I thought alter_table logic tests were failing and also some FK related tests in TestTenantLogic.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_opZJ8DGzJtp0xwG71:b-7xnpgj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)



<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
TestExecBuild_hash_sharded_index may have been one

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_ote8t1AqskXnxAK33:b-2zecih)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)



<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
I can't seem to access your commit via git, even after fetching from your fork. Could you try opening a draft PR for it and running the full suite of tests?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_paAIE-LaS8d54IUGF:b-w038jf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)



<!-- Sent from Reviewable.io -->

mgartner commented:
Here's a draft PR: https://github.com/cockroachdb/cockroach/pull/107305. All of the tests are passing.

msirek reviewed (commented):
Great, thanks!  This does look cleaner. I've updated my branch to use this fix.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_tO7o-9NPoE2UW7Y-s:bmc2q3b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @mgartner, and @michae2)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 7 files at r2, 1 of 8 files at r3, 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_u5kUl9T8B1Pb3PxyH:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, and @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I have a question, but it doesn't need to hold up merging.

> Reviewed 1 of 8 files at r3, 1 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_uAWTBETeQGwTcaHqQ:bip4jwh)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, and @msirek)

___
*[`pkg/sql/opt/optbuilder/select.go` line 867 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_u9vqU56RrG79wS9Ub:-N_u9vqU56RrG79wS9Uc:b-c6dlet) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f0d3ea3652499a12ef644e30489ad565ebe8d7c/pkg/sql/opt/optbuilder/select.go#L867)):*
> ```Go
> 			continue
> 		}
> 		if tabCol.IsMutation() && !includeVirtualMutationColOrds.Contains(i) {
> ```

I'm a little confused by this map. Would it be equivalent to write `if tabCol.IsMutation() && !tabCol.IsVirtualComputed()`? (In other words, does the map do anything? It seems like all virtual computed mutation columns are always in the map?)


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, and @michae2)

___
*[`pkg/sql/opt/optbuilder/select.go` line 867 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_u9vqU56RrG79wS9Ub:-N_uDpZ848p1gVepnVi6:b-ihc95s) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f0d3ea3652499a12ef644e30489ad565ebe8d7c/pkg/sql/opt/optbuilder/select.go#L867)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I'm a little confused by this map. Would it be equivalent to write `if tabCol.IsMutation() && !tabCol.IsVirtualComputed()`? (In other words, does the map do anything? It seems like all virtual computed mutation columns are always in the map?)
</blockquote></details>

I don't think it's equivalent. `virtualMutationColOrds` only contains those virtual computed columns present in the `ordinals` slice, which might be a subset of the columns.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
TFTRs!
r=mgartner,michae2

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-:-N_uEaX09k4V-lqiCr2K:b-ax2tej)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, and @michae2)



<!-- Sent from Reviewable.io -->

msirek commented:
bors r=mgartner,michae2

mgartner commented:
> I don't think it's equivalent. virtualMutationColOrds only contains those virtual computed columns present in the ordinals slice, which might be a subset of the columns.

Right. The alternative suggested by @michae2 would always include virtual mutation columns in `tabMeta.ComputedCols`. The current PR only includes virtual mutations columns when they exist in `ordinals` - which is when mutation columns were requested in the call to `tableOrdinals`.

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106782)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, and @msirek)

___
*[`pkg/sql/opt/optbuilder/select.go` line 867 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/106782#-N_u9vqU56RrG79wS9Ub:-N_uFiET-0c7TNSo2khY:bs7h64z) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f0d3ea3652499a12ef644e30489ad565ebe8d7c/pkg/sql/opt/optbuilder/select.go#L867)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

I don't think it's equivalent. `virtualMutationColOrds` only contains those virtual computed columns present in the `ordinals` slice, which might be a subset of the columns.
</blockquote></details>

Ah, I see! Thank you!


<!-- Sent from Reviewable.io -->
